pyccel/ast/utilities.py

```diff
@@ -53,7 +53,7 @@ def builtin_function(expr, args=None):
         return dic[name](*args)
 
     if name == 'map':
-        func = str(expr.args[0].name)
+        func = expr.args[0]
         args = [func]+list(args[1:])
         return PythonMap(*args)
```

Comment:
`args` is an argument to the function. It is not `expr.args`

Reply:
You're right, I read `expr.args[1:]` but there is no `expr` there

Reply:
Nevertheless, can't we write this with just one (clear!) line?

```python
if name == 'map':
    return PythonMap(expr.args[0], *args[1:])
```

By the way, I don't see why `args[0]` is not used here. If `args[0]` coincides with `expr.args[0]`, then my original comment was correct... Could you check please?

Reply:
`args[0]` contains a `FunctionDef` while `expr.args[0]` contains a `Symbol`. This is because it is the `FunctionDef` that is collected when we visit a `Symbol` (when we visit the argument). This should make sense, however the map is then usually used in a `For` loop. Personally I think this section is poorly written.

In a `For` block there are a few steps:
- Visit the iterable to get information about the type of iterable (map/zip/variable/etc)
- Construct a syntactic node representing the collection of the iterable at each pass of the loop from the visited iterable
    E.g. For a Variable the created node is
    ```
    assign = Assign(iterator, IndexedElement(iterable, indx))
    ```
    For me this step has a logical flaw as Syntactic objects are created from a Semantic object, despite the fact that the semantic stage does not work if you pass it a semantically treated object.
    For the map this is seen [here](https://github.com/pyccel/pyccel/blob/master/pyccel/parser/semantic.py#L2003-L2010) as we construct a `FunctionCall` from the arguments of the map. The line in `ast/headers.py` ensures that the first argument is a syntactic object and not a semantic object. This object will always cause a problem if it is a `FunctionDef` instead of a `Symbol` but the other arguments could too if they were more complicated than `Variable`
- Visit the newly constructed body

We can make the changes above by also changing [this line](https://github.com/pyccel/pyccel/blob/master/pyccel/parser/semantic.py#L2005) to `func   = iterable.args[0].name`

Ideally we would also ensure that all the arguments passed are syntactic objects, but I think this deserves a PR of its own to clean up `_visit_For`